core/scheduler/service/replay_worker.go

```diff
@@ -28,281 +36,229 @@ type ReplayScheduler interface {
-	if err := w.replayRepo.UpdateReplay(ctx, replayReq.Replay.ID(), state, updatedRuns, message); err != nil {
-		w.l.Error("unable to update replay with replay_id [%s]: %s", replayReq.Replay.ID().String(), err)
-		return err
-	}
-	raiseReplayMetric(replayReq.Replay.Tenant(), replayReq.Replay.JobName(), state)
-	return nil
+	return me.ToErr()
 }
 
-func identifyUpdatedRunStatus(existingJobRuns, incomingJobRuns []*scheduler.JobRunStatus) map[time.Time]scheduler.State {
+// syncStatus syncs existing and incoming runs
+// replay status: created -> in_progress -> [success, failed]
+// replay runs: [missing, pending] -> in_progress -> [success, failed]
+func syncStatus(existingJobRuns, incomingJobRuns []*scheduler.JobRunStatus) []*scheduler.JobRunStatus {
 	incomingRunStatusMap := scheduler.JobRunStatusList(incomingJobRuns).ToRunStatusMap()
-
-	updatedReplayMap := make(map[time.Time]scheduler.State)
-	for _, run := range existingJobRuns {
-		if run.State != scheduler.StateInProgress {
-			continue
-		}
-		if incomingRunStatusMap[run.ScheduledAt.UTC()] == scheduler.StateSuccess || incomingRunStatusMap[run.ScheduledAt.UTC()] == scheduler.StateFailed {
-			updatedReplayMap[run.ScheduledAt.UTC()] = incomingRunStatusMap[run.ScheduledAt.UTC()]
+	existingRunStatusMap := scheduler.JobRunStatusList(existingJobRuns).ToRunStatusMap()
+
+	updatedRunStatusMap := make(map[time.Time]scheduler.State)
+	for scheduledAt, existingState := range existingRunStatusMap {
+		switch existingState {
+		case scheduler.StateSuccess, scheduler.StateFailed:
+			updatedRunStatusMap[scheduledAt] = existingState
+		case scheduler.StateInProgress:
+			if incomingState, ok := incomingRunStatusMap[scheduledAt]; !ok {
+				updatedRunStatusMap[scheduledAt] = scheduler.StatePending
+			} else if incomingState == scheduler.StateSuccess || incomingState == scheduler.StateFailed {
+				updatedRunStatusMap[scheduledAt] = incomingState
+			} else {
+				updatedRunStatusMap[scheduledAt] = scheduler.StateInProgress
+			}
+		default: // pending state
+			if _, ok := incomingRunStatusMap[scheduledAt]; !ok {
+				updatedRunStatusMap[scheduledAt] = scheduler.StateMissing
+			} else {
+				updatedRunStatusMap[scheduledAt] = scheduler.StatePending
+			}
 		}
 	}
-	return updatedReplayMap
-}
 
-func (w ReplayWorker) fetchRun(ctx context.Context, replayReq *scheduler.ReplayWithRun, jobCron *cron.ScheduleSpec, scheduledAt time.Time) (*scheduler.JobRunStatus, error) {
-	jobRunCriteria := &scheduler.JobRunsCriteria{
-		Name:      replayReq.Replay.JobName().String(),
-		StartDate: scheduledAt,
-		EndDate:   scheduledAt,
-	}
-	runs, err := w.scheduler.GetJobRuns(ctx, replayReq.Replay.Tenant(), jobRunCriteria, jobCron)
-	if err != nil {
-		return nil, err
-	}
-	if len(runs) == 0 {
-		return nil, errors.NotFound(scheduler.EntityReplay, fmt.Sprintf("run on job %s with scheduled_at %s not exist", replayReq.Replay.JobName().String(), scheduledAt))
+	updatedJobRuns := []*scheduler.JobRunStatus{}
```

Comment:
let's use make instead, and use the index for assignment

Reply:
as discussed, decided to not use make as it will require us to get the index when iterating a map